sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -637,6 +690,54 @@ class kernel_bundle_impl {
     return isMangled ? Name : "__sycl_kernel_" + Name;
   }
 
+  std::string mangle_device_global_name(const std::string &Name) {
+    // TODO: Support device globals declared in namespaces.
+    return "_Z" + std::to_string(Name.length()) + Name;
+  }
+
+  bool is_valid_device(const device &DeviceCand) {
+    // Check if the device is in this bundle's list of devices.
+    if (std::count(MDevices.begin(), MDevices.end(), DeviceCand)) {
+      return true;
+    }
+
+    // Otherwise, if the device candidate is a sub-device it is also valid if
+    // its parent is valid.
+    if (!getSyclObjImpl(DeviceCand)->isRootDevice()) {
+      try {
+        return is_valid_device(
+            DeviceCand.get_info<info::device::parent_device>());
+      } catch (std::exception &e) {
+        __SYCL_REPORT_EXCEPTION_TO_STREAM("exception in is_valid_device", e);
```

Comment:
This `__SYCL_REPORT_EXCEPTION_TO_STREAM` macro is usually used in `noexcept` functions (like destructors) when we have no other option.  It doesn't seem needed here.  Why not let the exception go free so that the caller can decide? If your intention is to suppress it, then I'd recommend doing that and leaving a comment why. 

Also, that macro will `assert(false)` in ndef NDEBUG , which will bring the app to a halt.